src/network/network.cpp

```diff
@@ -7,201 +7,219 @@
+  for (auto const &node : conf_.network_boot_nodes) {
+    Public pub(node.id);
+    if (pub == key.pub()) {
+      LOG(log_wr_) << "not adding self to the boot node list";
       continue;
     }
+    LOG(log_nf_) << "Adding boot node:" << node.ip << ":" << node.tcp_port;
     auto ip = resolveHost(node.ip, node.tcp_port);
-    host_->addBootNode(dev::Public(node.id),
-                       dev::p2p::NodeIPEndpoint(ip.second.address(), node.udp_port, node.tcp_port));
-    boot_node_added++;
+    boot_nodes_[pub] = dev::p2p::NodeIPEndpoint(ip.second.address(), node.tcp_port, node.tcp_port);
   }
-  LOG(log_nf_) << " Number of boot node added: " << boot_node_added << std::endl;
-  host_->start(boot_node);
-  LOG(log_nf_) << "Started Network address: " << conf_.network_address << ":" << conf_.network_tcp_port << " :"
-               << conf_.network_udp_port << std::endl;
-  LOG(log_nf_) << "Started Node id: " << host_->id();
+  LOG(log_nf_) << " Number of boot node added: " << boot_nodes_.size() << std::endl;
+  string net_version = "TaraxaNode";  // TODO maybe give a proper name?
+  dev::bytes networkData;
+  if (!network_file_.empty()) {
+    networkData = contents(network_file_.string());
+  }
+  if (!networkData.empty()) {
+    host_ = std::make_shared<dev::p2p::Host>(net_version, net_conf, move(taraxa_net_conf), &networkData);
+  } else {
+    host_ = std::make_shared<dev::p2p::Host>(net_version, makeENR(key, net_conf), net_conf, move(taraxa_net_conf));
+  }
+  taraxa_capability_ = std::make_shared<TaraxaCapability>(*host_, tp_.unsafe_get_io_context(), conf_, db, pbft_mgr,
+                                                          pbft_chain, vote_mgr, next_votes_mgr, dag_mgr, dag_blk_mgr,
+                                                          trx_mgr, conf_.network_performance_log, key.address());
+  host_->registerCapability(taraxa_capability_);
 }
 
-void Network::stop() {
+Network::~Network() {
   if (bool b = false; !stopped_.compare_exchange_strong(b, !b)) {
     return;
   }
-
+  tp_.stop();
   host_->stop();
   if (!network_file_.empty()) {
-    saveNetwork(network_file_);
+    if (auto netData = host_->saveNetwork(); !netData.empty()) {
+      writeFile(network_file_.string(), &netData);
+      LOG(log_dg_) << "Network saved to: " << network_file_;
+    }
   }
-  taraxa_capability_->onStopping();
 }
 
-void Network::sendTest(NodeID const &id) {
-  taraxa_capability_->sendTestMessage(id, 1);
-  LOG(log_dg_) << "Sent test";
+void Network::start() {
+  if (bool b = true; !stopped_.compare_exchange_strong(b, !b)) {
+    return;
+  }
+  assert(!host_->isStarted());
+  host_->start();
+  if (!boot_nodes_.empty()) {
+    for (auto const &[k, v] : boot_nodes_) {
+      host_->addNode(k, v);
+    }
+    // Every 30 seconds check if connected to another node and refresh boot nodes
+    util::post_periodic(tp_.unsafe_get_io_context(), {30000}, [this] {
+      // If node count drops to zero add boot nodes again and retry
```

Comment:
post_periodic should be rather threadpool method and it would be obvious that it is processed by threadpool